Toshi/Source/TMemory.cpp

```diff
@@ -1,18 +1,283 @@
+				{
+					block = TakeoutAvailableBlock();
+					blockRegion->Block = block;
+					blockRegion->Block->Count = (size_t)alignedMemory;
+
+					// place taken block to the UsedBlocks container
+					block->Container = &m_UsedBlocksContainer;
+					block->Prev = &m_UsedBlocksContainer;
+					block->Next = m_UsedBlocksContainer.Next;
+					m_UsedBlocksContainer.Next->Prev = block;
+					m_UsedBlocksContainer.Next = block;
+
+					m_UsedBlocksContainer.Count++;
+				}
+
+				if (size != 0)
+				{
+					strncpy_s(blockRegion->Name, name, 52);
+					blockRegion->Size = alignedSize - sizeof(MemoryBlockRegion);
+
+					// place the first empty chunk right after blockRegion
+					MemoryChunk* chunk = (MemoryChunk*)((char*)blockRegion + sizeof(MemoryBlockRegion));
+					chunk->Region = blockRegion;
+					chunk->Next = nullptr;
+					chunk->Prev = nullptr;
+					chunk->Size = blockRegion->Size - sizeof(MemoryChunk);
+					chunk->Data = (char*)chunk + sizeof(MemoryChunk);
+
+					blockRegion->UsedChunks = nullptr;
+					blockRegion->FreeChunks = chunk;
+
+					result = blockRegion;
+				}
+
+				ReleaseMutex(s_Mutex);
+			}
+
+			ReleaseMutex(s_Mutex);
+			return result;
+		}
+
+		MemoryBlockInfo* TMemManager::TakeoutAvailableBlock()
+		{
+			assert(m_BlocksContainer.Count > 0 && "Ran out of Toshi memory blocks");
+			auto block = m_BlocksContainer.Next;
+
+			if (block != &m_BlocksContainer)
+			{
+				/*
+				*  Since we take the first block after the main one
+				*  block->Prev is always equal to m_BlocksContainer
+				*/
+
+				block->Container = 0;
+				block->Prev->Next = block->Next;
+				block->Next->Prev = block->Prev;
+
+				block->Next = block;
+				block->Prev = block;
+
+				m_BlocksContainer.Count--;
+
+				return block;
+			}
+
+			return nullptr;
+		}
+
+		static void AddChunkToList(MemoryChunk** list, MemoryChunk* chunk)
+		{
+			if (*list == nullptr)
+			{
+				*list = chunk;
+			}
+			else
+			{
+				chunk->Next = *list;
+				(*list)->Prev = chunk;
+				*list = chunk;
+			}
+		}
+
+		static void DeleteChunkFromList(MemoryChunk** list, MemoryChunk* chunk)
+		{
+			if (*list == chunk)
+			{
+				*list = chunk->Next;
+			}
+			else
+			{
+				chunk->Prev->Next = chunk->Next;
+				if (chunk->Next) { chunk->Next->Prev = chunk->Prev; }
+			}
+
+			chunk->Next = nullptr;
+			chunk->Prev = nullptr;
+		}
+
+		static void DefragmentChunks(MemoryChunk** list)
+		{
+			auto chunk1 = *list;
+			auto chunk2 = *list;
+
+			while (chunk1 != nullptr)
+			{
+				MemoryChunk* mergeableChunk = (MemoryChunk*)((char*)chunk1 + sizeof(MemoryChunk) + chunk1->Size);
+
+				// look for a mergeable chunk in the list
+				while (chunk2 != nullptr)
+				{
+					if (chunk2 == mergeableChunk)
+					{
+						chunk1->Size += chunk2->Size;
+						DeleteChunkFromList(list, chunk2);
+						break;
+					}
+
+					chunk2 = chunk2->Next;
+				}
+
+				chunk1 = chunk1->Next;
+			}
+		}
+	}
+
+	void* tmalloc(size_t size, TMemory::MemoryBlockRegion* block)
```

Comment:
Does that align with the actual game or are the parameters guesses?

Reply:
it aligns with the game